### PR TITLE
feat: allow dumping response buffer to some file

### PR DESCRIPTION
**Pull request by eggachecat**
@bogdanfinn Hi this is a PR for enabling saving in-process bytes to given-path-file.

With this feature one can easily implement some callback functions from their app (with other languages they just need to watch the file).

May you help to review this ?

### CODE REVIEW

**Comment by eggachecat** [comment]
background: I am trying to use [acheong08/ChatGPT](https://github.com/acheong08/ChatGPT) which depends on 
tls-client. And UI/UX is unacceptably slow because of no-stream-response-support

File: cffi_src/factory.go [hunk 1]
```diff
@@ -135,10 +136,54 @@ func BuildRequest(input RequestInput) (*http.Request, *TLSClientError) {
...
 
 // BuildResponse constructs a client response from a given HTTP response. The client response can then be sent to the interface consumer.
-func BuildResponse(sessionId string, withSession bool, resp *http.Response, cookies []*http.Cookie, isByteResponse bool) (Response, *TLSClientError) {
+func BuildResponse(sessionId string, withSession bool, resp *http.Response, cookies []*http.Cookie, input RequestInput) (Response, *TLSClientError) {
 	defer resp.Body.Close()
+	isByteResponse := input.IsByteResponse
+	var respBodyBytes []byte
+	var err error
+
+	if input.StreamOutputPath != nil {
```
Comment (bogdanfinn) [inline cffi_src/factory.go:145]: i would like to see the whole handling of writing the response to a file in a separate method.
would increase readability imho.

File: cffi_src/factory.go [hunk 2]
```diff
@@ -135,10 +136,54 @@ func BuildRequest(input RequestInput) (*http.Request, *TLSClientError) {
...
+			respBodyBytes = append(respBodyBytes, buf[:n]...)
+			if _, err = f.Write(buf[:n]); err != nil {
+				if input.WithDebug {
+					fmt.Printf("Append stream output error: %+v\n", err)
+				}
+				return Response{}, NewTLSClientError(err)
+			}
+
+			if input.WithDebug {
+				fmt.Printf("[stream decode result]==========\n%+v\n==========\n", string(buf[:n]))
```
Comment (bogdanfinn) [inline cffi_src/factory.go:176]: for now fine i guess. i will change that in a later PR to have the logger instance in those functions.

File: cffi_src/factory.go [hunk 3]
```diff
@@ -135,10 +136,54 @@ func BuildRequest(input RequestInput) (*http.Request, *TLSClientError) {
...
 // BuildResponse constructs a client response from a given HTTP response. The client response can then be sent to the interface consumer.
-func BuildResponse(sessionId string, withSession bool, resp *http.Response, cookies []*http.Cookie, isByteResponse bool) (Response, *TLSClientError) {
+func BuildResponse(sessionId string, withSession bool, resp *http.Response, cookies []*http.Cookie, input RequestInput) (Response, *TLSClientError) {
 	defer resp.Body.Close()
+	isByteResponse := input.IsByteResponse
+	var respBodyBytes []byte
+	var err error
+
+	if input.StreamOutputPath != nil {
+		f, err := os.OpenFile(*input.StreamOutputPath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
```
Comment (bogdanfinn) [inline cffi_src/factory.go:146]: just curious: did you test that under windows? 
as i'm only using mac i rarely test under windows :-D

File: cffi_src/factory.go [hunk 4]
```diff
@@ -135,10 +136,54 @@ func BuildRequest(input RequestInput) (*http.Request, *TLSClientError) {
...
+		if err != nil {
+			panic(err)
```
Comment (bogdanfinn) [inline cffi_src/factory.go:148]: don't panic here, but return a *TLSClientError

File: cffi_src/factory.go [hunk 5]
```diff
@@ -135,10 +136,54 @@ func BuildRequest(input RequestInput) (*http.Request, *TLSClientError) {
...
+					fmt.Printf("Append stream output error: %+v\n", err)
+				}
+				return Response{}, NewTLSClientError(err)
+			}
+
+			if input.WithDebug {
+				fmt.Printf("[stream decode result]==========\n%+v\n==========\n", string(buf[:n]))
+			}
+
+			if err != nil {
```
Comment (bogdanfinn) [inline cffi_src/factory.go:179]: this error check is redundant or am i wrong?

**Review by bogdanfinn — CHANGES_REQUESTED** [review]
Thank you for submitting a Pull Request.

I left some comments. Let me know what you think about that :-)

File: cffi_src/factory.go (hunk 1 above)
Comment (eggachecat) [reply cffi_src/factory.go]: Done! What do you think for the new version?

File: cffi_src/factory.go (hunk 2 above)
Comment (eggachecat) [reply cffi_src/factory.go]: Cool

File: cffi_src/factory.go (hunk 3 above)
Comment (eggachecat) [reply cffi_src/factory.go]: Yes I have tested it for the chatgpt program. It works just fine.

File: cffi_src/factory.go (hunk 5 above)
Comment (eggachecat) [reply cffi_src/factory.go]: Yes you are right. Have removed the redundance.

**Comment by bogdanfinn** [comment]
Thank you for contribution! @eggachecat

**Comment by acheong08** [comment]
So it streams normally or does it require file?

**Comment by eggachecat** [comment]
It requires files. Clients need to poll to update I believe

**Comment by eggachecat** [comment]
```python

def parse_file(file_name):
    try:
        with open(file_name, "rb") as fp:
            now = fp.read().decode()
            lines = [line for line in now.split('\n\n') if len(line) > 0]
            if len(lines) == 0:
                return None, False
            if len(lines) == 1:
                return lines[-1], False

            latest_resp_0 = lines[-1]
            latest_resp_1 = lines[-2]

            if "DONE" in latest_resp_0:
                print("detect done")
                return latest_resp_1, True

            latest_resp = latest_resp_0 if len(latest_resp_0) >= len(latest_resp_1) else latest_resp_1
            return latest_resp, False
    except Exception as e:
        print(traceback.format_exc())
        return None, False
    return None, False


def watch_resp_stream(file_name, callback, max_same_wait=10, sleep_ms=1):
    prev = None
    same_ctr = 0
    time_ctr = 0
    while same_ctr < max_same_wait or time_ctr > 300:
        latest, done = parse_file(file_name)
        if latest is not None:
            if done:
                callback(latest, True)
                break
            if latest == prev:
                same_ctr += 1
            else:
                callback(latest)
            prev = latest
        time.sleep(sleep_ms)
        time_ctr += 1
         def make_callback(scope):
            batch_text_size = 100
            lock = threading.Lock()
            has_been_last = False

            def _f(raw_text, is_last=False):
                print('processing with', raw_text)
                nonlocal scope, lock, has_been_last
                while lock.locked():
                    print('waiting lock...')
                    time.sleep(1)
                lock.acquire()
                print("scope['n']", scope['n'])
                try:
                    if "data:" in raw_text:
                        text = json.loads(raw_text.split("data:")[1])['message']['content']['parts'][0]
                    else:
                        text = raw_text

                    print("len(text)", len(text))
                    print("is_last", is_last)
                    print("has_been_last", has_been_last)
                    if len(text) > scope['n'] + batch_text_size or (is_last and not has_been_last):
                        msg = text[scope['n']:]
                        if scope['n'] != 0:
                            msg = "..." + msg
                        if not is_last:
                            msg += "..."
                        if is_last:
                            has_been_last = True
                        self.wb.send_txt_msg(to=to, content=msg)
                        scope['n'] = len(text)
                except Exception as e:
                    print(traceback.format_exc())
                lock.release()

            return _f

scope = {
     'n': 0,
}

 stream_callback  = make_callback(scope)
 t = threading.Thread(target=watch_resp_stream, args=(file_name, stream_callback, 5,))
 t.start()
```

Here's code for ChatGPT I am using.

**Comment by acheong08** [comment]
My solution was to write a proxy in Go and use that via normal requests from Python

**Comment by bogdanfinn** [comment]
> My solution was to write a proxy in Go and use that via normal requests from Python

Check out this repository maybe: https://github.com/bogdanfinn/tls-client-api

**Comment by GunGunGun** [comment]
Could I get an example of how to use this feature, in my case I want to use Python to read reponse from tls-client by chunks, for example `8192` bytes per while loop.

**Pull request merged**